src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs

```diff
@@ -1,462 +1,407 @@
+        public void DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsPathTooLongException()
         {
-            string path = IOServices.RemoveTrailingSlash(file.Path);
+            // While paths themselves can be up to 260 characters including trailing null, file systems
+            // limit each components of the path to a total of 255 characters.
+            var paths = IOInputs.GetPathsWithComponentLongerThanMaxComponent();
 
-            Assert.Throws<IOException>(() =>
+            Assert.All(paths, (path) =>
             {
-                Directory.CreateDirectory(path);
+                Assert.Throws<PathTooLongException>(() => Create(path));
             });
         }
-    }
-
 
-    [Fact]
-    public static void CreateDirectory_FileWithTrailingSlashAsPath_ThrowsIOException()
-    {
-        using (TemporaryFile file = new TemporaryFile())
+        [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
+        public void DirectoryLongerThanMaxPathAsPath_ThrowsPathTooLongException()
         {
-            string path = IOServices.AddTrailingSlashIfNeeded(file.Path);
-
-            Assert.Throws<IOException>(() =>
+            var paths = IOInputs.GetPathsLongerThanMaxPath();
+            Assert.All(paths, (path) =>
             {
-                Directory.CreateDirectory(path);
+                Assert.Throws<PathTooLongException>(() => Create(path));
             });
         }
-    }
 
-    [Fact]
-    public static void CreateDirectory_ExistingDirectoryWithoutTrailingSlashAsPath_DoesNotThrow()
-    {
-        using (TemporaryDirectory directory = new TemporaryDirectory())
+        [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
+        public void DirectoryLongerThanMaxDirectoryAsPath_ThrowsPathTooLongException()
         {
-            string path = IOServices.RemoveTrailingSlash(directory.Path);
-
-            DirectoryInfo result = Directory.CreateDirectory(path);
-
-            Assert.Equal(path, result.FullName);
-            Assert.True(Directory.Exists(result.FullName));
+            var paths = IOInputs.GetPathsLongerThanMaxDirectory();
+            Assert.All(paths, (path) =>
+            {
+                Assert.Throws<PathTooLongException>(() => Create(path));
+            });
         }
-    }
 
-    [Fact]
-    public static void CreateDirectory_ExistingDirectoryWithTrailingSlashAsPath_DoesNotThrow()
-    {
-        using (TemporaryDirectory directory = new TemporaryDirectory())
+        [Fact]
+        [PlatformSpecific(PlatformID.AnyUnix)]
+        [ActiveIssue(645)]
+        public void UnixLongPath_Allowed()
         {
-            string path = IOServices.AddTrailingSlashIfNeeded(directory.Path);
-
-            DirectoryInfo result = Directory.CreateDirectory(path);
-
-            Assert.Equal(path, result.FullName);
-            Assert.True(Directory.Exists(result.FullName));
+            var path = new string('a', 280);
+            Create(Path.Combine(TestDirectory, path));
```

Comment:
Yeah you're right. This needs to be split into two distinct tests - one to test maxcomponent fails on both platforms, one to test longpath passes on unix but fails on Windows.